### PR TITLE
[CALCITE-6464] Type inference for DECIMAL division seems incorrect

### PR DESCRIPTION
**Pull request by mihaibudiu**
Alternative PR to #3855. I have taken some tests from that PR, but adjusted the algorithm.
See the JIRA and the comments for a justification.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/type/RelDataTypeSystem.java [hunk 1]
```diff
@@ -294,21 +294,21 @@ default boolean shouldUseDoubleMultiplication(RelDataTypeFactory typeFactory,
...
+        int precision = d + scale;
+
+  // Rules from
+  // https://learn.microsoft.com/en-us/sql/t-sql/data-types/precision-scale-and-length-transact-sql
+        int bound = getMaxNumericPrecision() - 6;  // 32 in the MS documentation
+        if (precision <= bound) {
+          scale = Math.min(scale, getMaxNumericScale() - (precision - scale));
+        } else {
+          // precision > bound
+          scale = Math.min(6, scale);
```
Comment (suibianwanwank) [inline core/src/main/java/org/apache/calcite/rel/type/RelDataTypeSystem.java:308]: If MaxNumericScale < 6, is it possible here to exceed

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/rel/type/RelDataTypeSystem.java:308]: I will fix this case too, but, frankly, if a type system for DECIMAL does not have 6 digits of scale it's probably not very useful

Comment (suibianwanwank) [reply core/src/main/java/org/apache/calcite/rel/type/RelDataTypeSystem.java]: +1

**Comment by mihaibudiu** [comment]
Can someone review this PR? It fixes a genuine bug.

**Review by suibianwanwank — APPROVED** [review]
LGTM, I'm not sure if this is the most applicable for calcite, but it looks like sqlserver's is better than PG's

File: core/src/main/java/org/apache/calcite/rel/type/RelDataTypeSystem.java [hunk 2]
```diff
@@ -294,22 +294,41 @@ default boolean shouldUseDoubleMultiplication(RelDataTypeFactory typeFactory,
...
+
+  // Rules from
+  // https://learn.microsoft.com/en-us/sql/t-sql/data-types/precision-scale-and-length-transact-sql
+        // Rules reproduced here in case the web page goes away.
+        // Note that in SQL server getMaxNumericPrecision() == 38 and
+        // getMaxNumericScale() > 6.
+        // In multiplication and division operations, we need precision - scale places to store
+        // the integral part of the result. The scale might be reduced using the following rules:
+        //
+        // - The resulting scale is reduced to min(scale, 38 - (precision-scale))
```
Comment (NobiGo) [inline core/src/main/java/org/apache/calcite/rel/type/RelDataTypeSystem.java:310]: Do we need to add another tests for these rules?

**Review by NobiGo — APPROVED** [review]

**Comment by mihaibudiu** [comment]
> LGTM, I'm not sure if this is the most applicable for calcite, but it looks like sqlserver's is better than PG's

The type system of SqlServer (at least the version described in that article) is very different from the type system of Postgres. Postgres supports essentially arbitrary precision decimals, and negative scales. But computations with very high precision are very costly. So it's not easy to say which one is "better".

File: core/src/main/java/org/apache/calcite/rel/type/RelDataTypeSystem.java (hunk 2 above)
Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/rel/type/RelDataTypeSystem.java:310]: I think that the existing tests cover these changes - as a proof I had to change many types in the division result types in the validator and operator tests. I also added a test for a type system with lower precision, which motivated filing this bug in the first place.

**Review by caicancai — APPROVED** [review]
LGTM

**Pull request merged**